orttraining/orttraining/python/training/ort_triton/_sorted_graph.py

```diff
@@ -0,0 +1,210 @@
+                fields = [str(f[1]) for f in attr.ListFields()]
+                attributes.append(f"{fields[0]}:{fields[2]}={fields[1]}")
+            attributes_str = ",".join(attributes)
+            nodes.append(f"{node.op_type}[{attributes_str}]({inputs_str})->({outputs_str})")
+        nodes_str = ",".join(nodes)
+        return f"{graph_inputs_str}|{str(len(self._graph.output))}|{constants_str}|{nodes_str}"
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+    @property
+    def const_nodes(self) -> List[NodeProto]:
+        return self._const_nodes
+
+    @property
+    def sorted_nodes(self) -> List[NodeProto]:
+        return self._sorted_nodes
+
+    @property
+    def original_graph(self) -> GraphProto:
+        return self._graph
+
+    @property
+    def node_arg_infos(self) -> Dict[str, TensorInfo]:
+        return self._node_arg_infos
+
+    @property
+    def elementwise_graph_outputs(self) -> Set[str]:
+        return self._elementwise_graph_outputs
+
+    def _decompose(self):
+        dispatch = DecomposeDispatch()
+        pos = 0
+        # If a node is complex, decompose it and insert the decomposed nodes at the same position.
+        # All complex Ops are defined in DecomposeDispatch.
+        # It's possible that the decomposed nodes are also complex, so we need to do the decompose recursively.
+        # For example, decomposed nodes for "Softmax" contains "ReduceMean",
+        # which will be decomposed to "ReduceSum" and "Div" further.
+        while pos < len(self._sorted_nodes):
+            node = self._sorted_nodes[pos]
+            if node in dispatch:
+                new_nodes = dispatch(node, self._graph, node_arg_infos=self._node_arg_infos)
+                if len(new_nodes) != 1 or new_nodes[0] != node:
+                    new_nodes = topological_sort(node.input, new_nodes)
+                    self._sorted_nodes[pos : pos + 1] = new_nodes
+                    continue
+            if node.op_type == "Constant":
+                value_attr = get_attribute(node, "value")
+                self._node_arg_infos[node.output[0]] = TensorInfo(
+                    value_attr.data_type,
+                    list(to_numpy_array(value_attr).shape),
+                )
+            else:
+                input_infos = []
+                for input in node.input:
+                    input_infos.append(self._node_arg_infos[input])
+                output_infos = TypeAndShapeInfer.infer(node, input_infos, self._graph)
+                for idx, output in enumerate(node.output):
+                    self._node_arg_infos[output] = output_infos[idx]
+            pos += 1
+
+    # Save the ONNX graphs for debug purpose. The original ONNX graph is the subgraph from backend.
+    # The processed ONNX graph is the subgraph after decompose, it also contains the concrete shapes for each arg.
+    def save_onnx(self, file_path_prefix):
```

Comment:
What happens if a model exceeds the 2GB limit from protobuf?

Reply:
Normally huge protobuf means huge initializers, but this feature supports sub-graphs, which can have only scalar initializer for now, not possible to have such huge size.